File: textbook/python/04-agents/05_rag_research.py
# Copyright 2026 Terrene Foundation
# SPDX-License-Identifier: Apache-2.0
"""
# ════════════════════════════════════════════════════════════════════════
# Kailash SDK Textbook — Agents / RAG Research Agent
# ════════════════════════════════════════════════════════════════════════
# OBJECTIVE: Build a RAGResearchAgent for retrieval-augmented generation
# LEVEL: Intermediate
# PARITY: Python-only — no Rust equivalent
# VALIDATES: RAGResearchAgent, RAGConfig, RAGSignature, vector store
#            integration, document management, convergence detection
#
# Run: uv run python textbook/python/04-agents/05_rag_research.py
# ════════════════════════════════════════════════════════════════════════
"""
from __future__ import annotations

import os

from kaizen_agents.agents.specialized.rag_research import (
    RAGResearchAgent,
    RAGConfig,
    RAGSignature,
    SAMPLE_AI_DOCUMENTS,
)
from kaizen.core.base_agent import BaseAgent
from kaizen.strategies.multi_cycle import MultiCycleStrategy

model = os.environ.get("DEFAULT_LLM_MODEL", "claude-sonnet-4-20250514")

# ── 1. RAG pattern: Retrieve + Augment + Generate ─────────────────────
# RAG agents combine vector search with LLM generation:
#   1. RETRIEVE: Semantic search over a document corpus
#   2. AUGMENT: Inject relevant documents into the LLM context
#   3. GENERATE: LLM synthesizes an answer with source attribution
#
# RAGResearchAgent is an autonomous agent (MultiCycleStrategy) that
# can iteratively refine its search: query -> fetch -> analyze -> refine.

# ── 2. RAGSignature — retrieval-augmented output ──────────────────────

assert "query" in RAGSignature._signature_inputs

assert "answer" in RAGSignature._signature_outputs
assert "sources" in RAGSignature._signature_outputs
assert "confidence" in RAGSignature._signature_outputs
assert "relevant_excerpts" in RAGSignature._signature_outputs
assert (
    "tool_calls" in RAGSignature._signature_outputs
), "ADR-013: tool_calls for convergence (web_search, fetch_url, etc.)"

# ── 3. RAGConfig — retrieval-specific settings ────────────────────────

config = RAGConfig()

assert config.top_k_documents == 3, "Retrieve top 3 documents by default"
assert config.similarity_threshold == 0.3, "Minimum similarity score"
assert config.embedding_model == "all-MiniLM-L6-v2", "Default embedding model"
assert config.max_cycles == 15, "Research may need many cycles"
assert config.mcp_enabled is True, "Autonomous agents discover MCP tools"

# Custom retrieval configuration
custom_config = RAGConfig(
    top_k_documents=5,
    similarity_threshold=0.4,
    embedding_model="all-mpnet-base-v2",
    max_cycles=20,
)
assert custom_config.top_k_documents == 5
assert custom_config.similarity_threshold == 0.4

# ── 4. Sample documents (built-in knowledge base) ─────────────────────
# RAGResearchAgent ships with sample AI documents for testing.
# In production, you provide your own document corpus.

assert len(SAMPLE_AI_DOCUMENTS) == 5

doc_ids = {doc["id"] for doc in SAMPLE_AI_DOCUMENTS}
assert "doc1" in doc_ids  # Machine Learning
assert "doc2" in doc_ids  # Deep Learning
assert "doc3" in doc_ids  # NLP
assert "doc4" in doc_ids  # Computer Vision
assert "doc5" in doc_ids  # Reinforcement Learning

# Every document has id, title, content
for doc in SAMPLE_AI_DOCUMENTS:
    assert "id" in doc
    assert "title" in doc
    assert "content" in doc
    assert len(doc["content"]) > 50, "Documents have substantial content"

# ── 5. RAGResearchAgent instantiation ─────────────────────────────────
# The agent initializes a vector store with sample documents
# and uses MultiCycleStrategy for autonomous execution.

agent = RAGResearchAgent(
    llm_provider="mock",
    model=model,
)

assert isinstance(agent, RAGResearchAgent)
assert isinstance(agent, BaseAgent)
assert isinstance(
    agent.strategy, MultiCycleStrategy
), "RAG MUST use MultiCycleStrategy for autonomous research"

# ── 6. Vector store — document management ─────────────────────────────
# The agent manages a vector store for semantic search.
# Documents are added, searched, and can be cleared.

assert agent.get_document_count() == 5, "Sample docs loaded by default"

# Add a custom document
agent.add_document(
    doc_id="doc6",
    title="Quantum Computing",
    content="Quantum computing uses quantum mechanics principles such as "
    "superposition and entanglement to perform computations that would be "
    "intractable for classical computers.",
)
assert agent.get_document_count() == 6

# Clear all documents
agent.clear_documents()
assert agent.get_document_count() == 0

# ── 7. Convergence detection for research ─────────────────────────────
# RAG uses the same ADR-013 objective convergence as ReAct:
# tool_calls non-empty -> continue researching
# tool_calls empty     -> research complete

# Research in progress (needs more sources)
result_researching = {
    "tool_calls": [{"name": "web_search", "params": {"query": "deep learning"}}],
}
assert agent._check_convergence(result_researching) is False

# Research complete (no more tools needed)
result_done = {"tool_calls": [], "confidence": 0.9}
assert agent._check_convergence(result_done) is True

# Malformed tool_calls -> stop for safety
result_malformed = {"tool_calls": "not a list"}
assert agent._check_convergence(result_malformed) is True

# Subjective fallback: high confidence + comprehensive depth
result_subjective = {"confidence": 0.9, "research_depth": "comprehensive"}
assert agent._check_convergence(result_subjective) is True

# Default: converged (safe fallback)
assert agent._check_convergence({}) is True

# ── 8. Input validation ───────────────────────────────────────────────

empty_result = agent.run(query="")
assert empty_result["error"] == "INVALID_INPUT"
assert empty_result["sources"] == []
assert empty_result["confidence"] == 0.0

# ── 9. Return structure ───────────────────────────────────────────────
# RAGResearchAgent.run() returns:
# {
#     "answer": "Machine learning is a subset of AI...",
#     "sources": ["doc1", "doc2"],
#     "confidence": 0.85,
#     "relevant_excerpts": [
#         {
#             "title": "Introduction to Machine Learning",
#             "excerpt": "Machine learning is a subset of...",
#             "similarity": 0.87,
#         }
#     ],
#     "retrieval_quality": 0.82,  # average similarity
# }
#
# NOTE: Full run() requires an LLM API key and a populated vector
# store. The above documents the return format.

# ── 10. Custom vector store ───────────────────────────────────────────
# For production use, you can provide your own vector store with
# domain-specific documents instead of the built-in samples.
#
# from kaizen.retrieval.vector_store import SimpleVectorStore
#
# store = SimpleVectorStore(embedding_model="all-mpnet-base-v2")
# store.add_documents([
#     {"id": "d1", "title": "My Doc", "content": "..."},
# ])
#
# agent = RAGResearchAgent(
#     model=model,
#     vector_store=store,
#     top_k_documents=5,
# )

# ── 11. Node metadata ─────────────────────────────────────────────────

assert RAGResearchAgent.metadata.name == "RAGResearchAgent"
assert "rag" in RAGResearchAgent.metadata.tags
assert "vector-search" in RAGResearchAgent.metadata.tags
assert "semantic" in RAGResearchAgent.metadata.tags

print("PASS: 04-agents/05_rag_research")

File: textbook/rust/04-agents/src/bin/05_rag_research.rs
// Copyright 2026 Terrene Foundation
// SPDX-License-Identifier: Apache-2.0
//! Kailash SDK Textbook -- Agents / RAG Research Agent
//!
//! OBJECTIVE: Build a retrieval-augmented generation agent for research tasks.
//! LEVEL: Intermediate
//! PARITY: Equivalent -- Python has RAGResearchAgent;
//!         Rust uses the same pattern with retriever tool integration.
//! VALIDATES: RAG pattern, retriever tools, context injection, citation tracking
//!
//! Run: cargo run -p tutorial-agents --bin 05_rag_research

use kailash_kaizen::agent::{
    InputField, OutputField, SignatureBuilder,
};
use kaizen_agents::delegate_engine::ToolDef;

fn main() {
    // ── 1. RAG pattern ──
    // Retrieval-Augmented Generation adds external knowledge to LLM calls:
    //   1. RETRIEVE: Fetch relevant documents from a knowledge base
    //   2. AUGMENT: Inject retrieved context into the prompt
    //   3. GENERATE: LLM produces an answer grounded in the context
    //
    // This prevents hallucination by grounding responses in real data.

    // ── 2. RAG Signature ──
    // The signature captures the full RAG interaction.

    let rag_sig = SignatureBuilder::new("RAGResearch")
        .description(
            "You are a research assistant. Answer questions using only \
             the provided context. Cite your sources. If the context \
             doesn't contain the answer, say so."
        )
        .input(InputField::new("question", "Research question"))
        .input(InputField::new("context", "Retrieved documents").with_default(""))
        .output(OutputField::new("answer", "Answer grounded in the context"))
        .output(OutputField::new("sources", "Sources used (document IDs or titles)"))
        .output(OutputField::new(
            "confidence",
            "Confidence that the answer is fully supported by context",
        ))
        .build();

    assert_eq!(rag_sig.inputs().len(), 2);
    assert_eq!(rag_sig.outputs().len(), 3);

    // ── 3. Retriever tool ──
    // The retriever is a tool the agent calls to fetch relevant documents.
    // It takes a query and returns ranked passages.

    let retriever_tool = ToolDef::new(
        "search_knowledge_base",
        "Search the knowledge base for relevant documents",
        serde_json::json!({
            "type": "object",
            "properties": {
                "query": {
                    "type": "string",
                    "description": "Search query for the knowledge base"
                },
                "top_k": {
                    "type": "integer",
                    "description": "Number of documents to retrieve",
                    "default": 5
                }
            },
            "required": ["query"]
        }),
    );

    assert_eq!(retriever_tool.name(), "search_knowledge_base");

    // ── 4. Additional RAG tools ──
    // Production RAG systems have multiple retrieval strategies.

    let _doc_lookup = ToolDef::new(
        "get_document",
        "Retrieve a specific document by ID",
        serde_json::json!({
            "type": "object",
            "properties": {
                "document_id": {"type": "string", "description": "Document ID"}
            },
            "required": ["document_id"]
        }),
    );

    let _web_search = ToolDef::new(
        "web_search",
        "Search the web for recent information",
        serde_json::json!({
            "type": "object",
            "properties": {
                "query": {"type": "string", "description": "Web search query"},
                "max_results": {"type": "integer", "default": 3}
            },
            "required": ["query"]
        }),
    );

    // ── 5. Multi-source RAG ──
    // Advanced RAG agents search multiple sources and cross-reference.

    let multi_rag_sig = SignatureBuilder::new("MultiSourceRAG")
        .description(
            "Research across multiple knowledge bases. Cross-reference \
             findings. Flag contradictions between sources."
        )
        .input(InputField::new("question", "Research question"))
        .output(OutputField::new("answer", "Synthesized answer from all sources"))
        .output(OutputField::new("sources", "All sources used with relevance scores"))
        .output(OutputField::new(
            "contradictions",
            "Any contradictions found between sources",
        ))
        .output(OutputField::new("gaps", "Information gaps identified"))
        .build();

    assert_eq!(multi_rag_sig.outputs().len(), 4);

    // ── 6. RAG execution pattern ──
    // The agent autonomously decides when to retrieve:
    //
    //   // 1. Agent receives question
    //   // 2. Agent calls search_knowledge_base tool
    //   // 3. Agent receives retrieved context
    //   // 4. Agent generates answer citing the context
    //   // 5. If confidence is low, agent retrieves more
    //
    //   let agent = RAGAgent::new(rag_sig, tools, config);
    //   let result = agent.run(inputs!{
    //       "question" => "What is the Terrene Foundation?"
    //   }).await;
    //
    // NOTE: We do not call run() here as it requires an LLM API key.

    // ── 7. Key concepts ──
    // - RAG: Retrieve, Augment, Generate -- grounds LLM in real data
    // - Retriever tool: agent-callable search over knowledge bases
    // - Citation tracking: sources output field for traceability
    // - Confidence: self-assessment of context coverage
    // - Multi-source: cross-referencing, contradiction detection
    // - Agent autonomously decides when and what to retrieve

    println!("PASS: 04-agents/05_rag_research");
}
